_posts/2020-08-17-creating-your-ci-cd-pipeline-with-concourse.md

```diff
@@ -0,0 +1,219 @@
+example could be software with a microservice architecture pattern. With this approach, you have a shared pipeline definition
+for all of your services.
+
+In my example, project variables are defined in files on local disk with a prefix *runtime_variables*.
+
+```yaml
+git-source: https://github.com/solita-alperttiti/concourse-ci-example.git
+git-branch: master
+``` 
+
+So, with these values my git resource is tied in a specific branch in a specific git repository.
+
+Variables are loaded into the Concourse when the pipeline is created. It is also possible to
+define multiple configuration files if you wish to split them up.
+
+Be noted, that variables can be loaded out from the Concourse afterward by using the Fly client with proper access,
+so in real-world cases it might be useful to consider using some credentials' manager if you have some sensitive data
+in your variables. [Concourse supports many different credential managers out-of-the-box](https://concourse-ci.org/creds.html).
+By using an external manager you only need to provide secrets required by the manager, when you are creating your pipeline.
+In general, when you are going to set up a new CI-system, you should always do a very careful evaluation from the
+security point of view. However, this is out of the scope of this blog post.
+
+
+Right, now we have some input. The goal in this example is to get some input from git,
+make some *things*, and then push the result (or output) into a specific repository in Docker Hub.
+So, next we define the output. In general, the output can be whatever is needed,
+e.g. rsync your project artifacts to some server, push something to git and so on.
+
+```yaml
+- name: concourse-example-registry
+  type: registry-image-resource
+  source:
+    repository: ((docker-source))
+    tag: ((docker-tag))
+    username: ((docker-username))
+    password: ((docker-password))
+```
+
+So, now we have external input and output. The next phase is to define a job that uses those. Let's use
+our imagination and call our job *build-and-publish*. You could define as many jobs as you wish,
+but in this case, one is enough.
+
+```yaml
+jobs:
+- name: build-and-publish
+```
+
+Every job has some series of steps which is called a *plan*.
+
+```yaml
+plan:
+    - get: concourse-example
+      trigger: ((trigger-build))
+    - task: prepare-build
+      file: concourse-example/concourse/pipelines/concourse-example/concourse-example-prepare-build.yml
+      vars:
+        docker-file: ((docker-file))
+        docker-tag: ((docker-tag))
+    - task: build
+      file: concourse-example/concourse/pipelines/concourse-example/concourse-example-build.yml
+      privileged: true
+      params:
+        DOCKERFILE: prebuild-output/((docker-file))
+    - put: concourse-example-registry
+      params: {
+        image: image/image.tar
+      }
+```
+
+So, while this definition is very self-explanatory let's walk it through anyway. First, we fetch our input
```

Comment:
Anyway could be dropped

Reply:
On the first draft I did not have this, but added it later. I want to point out that, it's might not be necessary go it through, but to be thorough I'll do it anyway.